File: java/markdown/src/main/java/Markdown.java
// Refactor by separated complex ul logic

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.ListIterator;
import java.util.stream.Collectors;

class Markdown {
  
    String parse(String markdown)
    {
        List<String> results = Arrays.stream(markdown.split("\n"))
            .map(l -> parseLine(l))
            .collect(Collectors.toList());

        return String.join("", addUl(results));
    }

    private String parseLine(String line)
    {
        if (line.startsWith("#")) {
            line = parseHeader(line);
        } else if (line.startsWith("*")) {
            line = parseListItem(line);
        } else {
            line = parseParagraph(line);
        }
        return parseSomeSymbols(line);
    }

    private List<String> addUl(List<String> lines)
    {
        boolean activeList = false;
        ListIterator<String> iterator = lines.listIterator();
        while (iterator.hasNext()) {
            String line = iterator.next();
            if (line.startsWith("<li>")) {
                if (!activeList) {
                    iterator.remove();
                    iterator.add("<ul>");
                    iterator.add(line);
                    activeList = true;
                }
            } else if (activeList) {
                iterator.remove();
                iterator.add("</ul>");
                iterator.add(line);
                activeList = false;
            }
        }

        if (activeList) {
            iterator.add("</ul>");
        }
        return lines;
    }

    private String parseHeader(String markdown)
    {
        int count = 0;
        while (markdown.charAt(count) == '#') {
            count++;
        }

        if (count == 0) {
            return null;
        }
        return String.format("<h%d>%s</h%d>", count, markdown.substring(count + 1), count);
    }

    private String parseListItem(String markdown) {
        if (markdown.startsWith("*")) {
            String listItemString = markdown.substring(2);
            return "<li>" + listItemString + "</li>";
        }

        return null;
    }

    private String parseParagraph(String markdown) {
        return "<p>" + markdown + "</p>";
    }

    private String parseSomeSymbols(String markdown) {
        return markdown.replaceAll("__(.+)__", "<strong>$1</strong>").replaceAll("_(.+)_", "<em>$1</em>");
    }
}

File: php/markdown/markdown.php
<?php

function parseMarkdown($markdown)
{
    $lines = array_map(
        function ($line) { return parseLine($line); },
        explode("\n", $markdown)
    );

    return implode("", addUl($lines));
}

function parseLine($line)
{
    list($match, $line) = parseHeader($line);
    if (!$match) {
        list($match, $line) = parseListItem($line);
    }
    if (!$match) {
        $line = "<p>{$line}</p>";
    }
    return parsePartial($line);
}

function parsePartial($line)
{
    $line = preg_replace("/__(.+)__/", "<em>$1</em>", $line);
    $line = preg_replace("/_(.+)_/", "<i>$1</i>", $line);
    return $line;
}

function parseHeader($line)
{
    if (!preg_match("/^#/", $line)) {
     return [false, $line];
    }

    $count = 0;
    foreach (str_split($line) as $char) {
        if ($char === '#') {
            $count++;
        } else {
            break;
        }
    }
    return [true, sprintf("<h%d>%s</h%d>", $count, substr($line, $count + 1), $count)];
}

function parseListItem($line)
{
    if (!preg_match("/^\*/", $line)) {
        return [false, $line];
    }

    $line = substr($line, 2);
    // adjust for strange test case
    if (preg_match("/^_/", $line)) {
        return [true,  "<li>" . $line . "</li>"];
    } else {
        return [true,  "<li><p>" . $line . "</p></li>"];
    }

}

function addUl($lines)
{
    $result = [];
    $inList = false;
    foreach ($lines as $line) {
        if (preg_match("/^<li>/", $line)) {
            if (!$inList) {
                $result[] = "<ul>";
                $inList = true;
            }
        } else {
            if ($inList) {
                $result[] = "</ul>";
                $inList = false;
            }
        }
        $result[] = $line;
    }

    if ($inList) {
        $result[] = "</ul>";
    }
    return $result;
}